PS-5206 : [PS8QA] handle_fatal_signal (sig=11) in handler::get_ha_sha…

satya-bodapati opened the pull request:
…re_ptr

Problem:
-------
Server crashes when ha_tokupart::clone() interface is used.

Fix:
----
To clone native partition tokupart handler, we will not use clone()
on each partition handler. This is because tokudb_create_handler() gives
ha_tokupart handler instead of ha_tokudb handlers. This happens becuase of
presence of parition info in TABLE_SHARE

We create a new partition handler in clone() with part_info from the current
handler.

satya-bodapati commented:
```
./mtr --mem --suite=tokudb.parts --parallel=4 --retry-failure=0|tee mtr.log
Logging: /home/satya/WORK/PS-5.7-5206/mysql-test/mysql-test-run.pl  --mem --suite=tokudb.parts --parallel=4 --retry-failure=0
2019-03-20T06:48:30.242327Z 0 [Warning] Changed limits: max_open_files: 1024 (requested 5000)
2019-03-20T06:48:30.242372Z 0 [Warning] Changed limits: table_open_cache: 431 (requested 2000)
MySQL Version 5.7.23
Checking supported features...
 - SSL connections supported
 - binaries are debug compiled
Using suites: tokudb.parts
Collecting tests...
 - adding combinations for tokudb.parts
Checking leftover processes...
Removing old var directory...
Creating var directory '/home/satya/WORK/PS-5.7-5206/bld/mysql-test/var'...
 - symlinking 'var' to '/dev/shm/var_auto_RPNs'
Installing system database...
Using parallel: 4

==============================================================================

TEST                                  WORKER RESULT   TIME (ms) or COMMENT
--------------------------------------------------------------------------

worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 13000..13009
worker[2] Using MTR_BUILD_THREAD 301, with reserved ports 13010..13019
worker[3] Using MTR_BUILD_THREAD 302, with reserved ports 13020..13029
worker[4] Using MTR_BUILD_THREAD 303, with reserved ports 13030..13039
tokudb.parts.part_supported_sql_func_tokudb [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter1_1_2_tokudb [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter1_1_tokudb   [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter1_2_tokudb   [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter2_1_1_tokudb [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter2_1_2_tokudb [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter2_2_1_tokudb [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter2_2_2_tokudb [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_alter4_tokudb     [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_basic_tokudb      [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_debug_tokudb      [ skipped ]  Test needs 'big-test' option
tokudb.parts.partition_basic_symlink_tokudb [ disabled ]  tokudb_file_per_table is not supported
tokudb.parts.partition_reorganize_tokudb [ disabled ]  tokudb_file_per_table is not supported
tokudb.parts.partition_debug_sync_tokudb 'native-partitioning' w2 [ pass ]   1099
tokudb.parts.partition_debug_sync_tokudb 'non-native-partitioning' w4 [ pass ]   1090
tokudb.parts.partition_mgm_lc2_tokudb 'non-native-partitioning' w4 [ skipped ]  Test requires: 'lowercase2'
tokudb.parts.partition_mgm_lc1_tokudb 'non-native-partitioning' w2 [ pass ]    465
tokudb.parts.partition_max_parts_hash_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_inv_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_key_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_list_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.disable_native_partitioning 'non-native-partitioning' w2 [ skipped ]  Test requires native partitioning enabled
tokudb.parts.partition_max_parts_range_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_key_list_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.nonflushing_analyze_debug 'non-native-partitioning' w2 [ pass ]     34
tokudb.parts.partition_max_sub_parts_key_range_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_list_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_range_tokudb 'non-native-partitioning' w4 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.part_blocked_sql_func_tokudb 'non-native-partitioning' w2 [ pass ]    928
tokudb.parts.partition_alter3_tokudb 'non-native-partitioning' w2 [ pass ]    663
tokudb.parts.partition_auto_increment_tokudb 'non-native-partitioning' w2 [ pass ]    434
tokudb.parts.partition_bit_tokudb 'non-native-partitioning' w2 [ pass ]    157
tokudb.parts.partition_char_tokudb 'non-native-partitioning' w2 [ pass ]    516
tokudb.parts.partition_datetime_tokudb 'non-native-partitioning' w2 [ pass ]    532
tokudb.parts.partition_decimal_tokudb 'non-native-partitioning' w2 [ pass ]    605
tokudb.parts.partition_engine_tokudb 'non-native-partitioning' w2 [ pass ]   1639
tokudb.parts.partition_exch_myisam_tokudb 'non-native-partitioning' w2 [ pass ]     69
tokudb.parts.partition_exch_qa_1_tokudb 'non-native-partitioning' w2 [ pass ]    161
tokudb.parts.partition_exch_qa_4_tokudb 'non-native-partitioning' w2 [ pass ]    244
tokudb.parts.partition_exch_qa_5_tokudb 'non-native-partitioning' w2 [ pass ]    160
tokudb.parts.partition_exch_qa_7_tokudb 'non-native-partitioning' w2 [ pass ]    138
tokudb.parts.partition_exch_qa_8_tokudb 'non-native-partitioning' w2 [ pass ]    256
tokudb.parts.partition_exch_tokudb 'non-native-partitioning' w2 [ pass ]    336
tokudb.parts.partition_special_tokudb 'native-partitioning' w3 [ pass ]  27267
tokudb.parts.partition_special_tokudb 'non-native-partitioning' w1 [ pass ]  27249
tokudb.parts.partition_mgm_lc1_tokudb 'native-partitioning' w2 [ pass ]    434
tokudb.parts.partition_mgm_lc2_tokudb 'native-partitioning' w3 [ skipped ]  Test requires: 'lowercase2'
tokudb.parts.partition_max_parts_hash_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_inv_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_key_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_list_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_parts_range_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_key_list_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_key_range_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_list_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_max_sub_parts_range_tokudb 'native-partitioning' w1 [ skipped ]  Need open_files_limit >= 24576 for myisam or 8192 for innodb (see ulimit -n) 
tokudb.parts.partition_decimal_tokudb 'native-partitioning' w3 [ pass ]    723
tokudb.parts.partition_engine_tokudb 'native-partitioning' w3 [ pass ]   1664
tokudb.parts.partition_exch_qa_7_tokudb 'native-partitioning' w1 [ pass ]    141
tokudb.parts.partition_exch_myisam_tokudb 'native-partitioning' w3 [ pass ]     91
tokudb.parts.partition_exch_qa_8_tokudb 'native-partitioning' w1 [ pass ]    289
tokudb.parts.partition_exch_qa_1_tokudb 'native-partitioning' w3 [ pass ]    214
tokudb.parts.partition_exch_tokudb 'native-partitioning' w1 [ pass ]    378
tokudb.parts.partition_exch_qa_4_tokudb 'native-partitioning' w3 [ pass ]    297
tokudb.parts.partition_exch_qa_5_tokudb 'native-partitioning' w3 [ pass ]    177
tokudb.parts.disable_native_partitioning 'native-partitioning' w2 [ pass ]   7376
tokudb.parts.partition_handler_clone 'native-partitioning' w3 [ pass ]     19
tokudb.parts.partition_int_tokudb 'native-partitioning' w3 [ pass ]   1191
tokudb.parts.partition_mgm_lc0_tokudb 'native-partitioning' w3 [ pass ]    506
tokudb.parts.nonflushing_analyze_debug 'native-partitioning' w2 [ pass ]     83
tokudb.parts.partition_mgm_lc10_tokudb 'native-partitioning' w3 [ pass ]    607
tokudb.parts.partition_syntax_tokudb 'native-partitioning' w3 [ pass ]    172
tokudb.parts.partition_value_tokudb 'native-partitioning' w3 [ skipped ]  # CAST() in partitioning function is currently not supported.
tokudb.parts.part_blocked_sql_func_tokudb 'native-partitioning' w2 [ pass ]   1024
tokudb.parts.partition_alter3_tokudb 'native-partitioning' w2 [ pass ]    630
tokudb.parts.partition_auto_increment_tokudb 'native-partitioning' w2 [ pass ]    391
tokudb.parts.partition_bit_tokudb 'native-partitioning' w2 [ pass ]    149
tokudb.parts.partition_char_tokudb 'native-partitioning' w2 [ pass ]    496
tokudb.parts.partition_datetime_tokudb 'native-partitioning' w2 [ pass ]    553
tokudb.parts.partition_exchange_tokudb 'non-native-partitioning' w4 [ pass ]  40154
tokudb.parts.partition_handler_clone 'non-native-partitioning' w4 [ skipped ]  Test requires native partitioning enabled
tokudb.parts.partition_float_tokudb 'non-native-partitioning' w2 [ pass ]   1280
tokudb.parts.partition_int_tokudb 'non-native-partitioning' w4 [ pass ]   1061
tokudb.parts.partition_mgm_lc0_tokudb 'non-native-partitioning' w2 [ pass ]    500
tokudb.parts.partition_syntax_tokudb 'non-native-partitioning' w2 [ pass ]    179
tokudb.parts.partition_mgm_lc10_tokudb 'non-native-partitioning' w4 [ pass ]    456
tokudb.parts.partition_value_tokudb 'non-native-partitioning' w2 [ skipped ]  # CAST() in partitioning function is currently not supported.
tokudb.parts.upgrade_from_prev_ver 'non-native-partitioning' w4 [ skipped ]  Test requires native partitioning enabled
tokudb.parts.partition_float_tokudb 'native-partitioning' w2 [ pass ]   1223
tokudb.parts.upgrade_from_prev_ver 'native-partitioning' w3 [ pass ]  13330
tokudb.parts.partition_exchange_tokudb 'native-partitioning' w1 [ pass ]  40162
--------------------------------------------------------------------------
The servers were restarted 14 times
Spent 179.992 of 90 seconds executing testcases

Completed: All 53 tests were successful.

36 tests were skipped, 25 by the test itself.
```

laurynas-biveinis commented on mysql-test/suite/tokudb.parts/t/partition_handler_clone.test:
```diff
@@ -0,0 +1,14 @@
+source include/have_tokudb.inc;
+--source ../include/have_native_partitioning.inc
+
+set default_storage_engine='tokudb';
```
Redundant

laurynas-biveinis commented on storage/tokudb/ha_tokupart.cc:
```diff
@@ -67,6 +67,13 @@ handler *ha_tokupart::clone(const char *name, MEM_ROOT *mem_root) {
 
+    /* We will not use clone() interface to clone individual partition
+    handlers. This is because tokudb_create_handler() gives ha_tokupart handler
+    instead of ha_tokudb handlers. This happens becuase of presence of parition
```
```suggestion
    instead of ha_tokudb handlers. This happens because of presence of partition
```

satya-bodapati commented:
>     * Consider creating a new constructor (to replace the old about-to-be-removed cloning constructor) which does the work of set_part_info/set_ha_share_ref calls below: it seems that it would be similar to the cloning constructor except it would not set m_clone_of.

done 

> 
>     * Is m_pkey_is_clustered set correctly after the patch, if it matters?
doesn't matter.
> 
>     * Please cleanup dead Partition_base::m_clone_of etc code in a follow-up. (If I understand correctly, Toku was its sole user)
No, rockspart is also its user. So when I removed this, rocksdb native paritioning is affected too.
I went for the whole changes and fixed the issue in rocksdb too.

satya-bodapati commented:
ready for review again (this time with the same fix ported to rocksdb as well).
Added rocksdb testcase

satya-bodapati commented:
https://ps57.cd.percona.com/job/percona-server-5.7-pipeline/284/
all rocksdb and tokudb suites

laurynas-biveinis commented on storage/rocksdb/ha_rockspart.cc:
```diff
@@ -97,6 +88,13 @@ handler *ha_rockspart::clone(const char *name, MEM_ROOT *mem_root) {
+  instead of ha_tokudb handlers. This happens because of presence of parition
+  info in TABLE_SHARE. New partition handlers are created for each partiton
+  in native_part::Partition_base::open() */
+  new_handler->get_partition_handler()->set_part_info(m_part_info, true);
```
Could this part be moved to ctor as well?

satya-bodapati replied:
yes, this is possible.

laurynas-biveinis commented on storage/rocksdb/ha_rockspart.cc:
```diff
@@ -97,6 +88,13 @@ handler *ha_rockspart::clone(const char *name, MEM_ROOT *mem_root) {
             (uchar *)alloc_root(mem_root, ALIGN_SIZE(ref_length) * 2)))
     goto err;
 
+  /* We will not use clone() interface to clone individual partition
```
With m_is_clone_of gone, is this comment still relevant? Isn't the current implementation the natural one now?

satya-bodapati replied:
Well, the natural instinct to clone ha_tokupart/rockspart/Partition_base::clone() is to again use clone() interface for individual partition handlers.

If it is not confusing, I would suggest to keep as is. But am OK with removal as well

laurynas-biveinis replied:
OK, let's keep it

laurynas-biveinis commented on storage/tokudb/ha_tokupart.cc:
```diff
@@ -67,6 +59,13 @@ handler *ha_tokupart::clone(const char *name, MEM_ROOT *mem_root) {
               (uchar *)alloc_root(mem_root, ALIGN_SIZE(ref_length) * 2)))
         goto err;
 
+    /* We will not use clone() interface to clone individual partition
```
Same two comments/questions as for Rocks

satya-bodapati commented:
review comments addressed. Ready for review again @laurynas-biveinis

laurynas-biveinis commented on storage/rocksdb/ha_rockspart.h:
```diff
@@ -24,12 +24,11 @@ class ha_rockspart : public native_part::Partition_base {
+  ha_rockspart(handlerton *hton, TABLE_SHARE *table_arg,
+               partition_info *part_info)
+      : native_part::Partition_base(hton, table_arg, &table_arg->ha_share) {
+    this->get_partition_handler()->set_part_info(part_info, true);
```
Shouldn't part_info be passed to Partition_base constructor, like the removed clone constructor did?

satya-bodapati replied:
set_ha_share_ref() is overrided by Partition_base and made private. Hence this has to be passed to Partition_base. If we can make it protected, we don't need to pass this to Partition_base.

laurynas-biveinis commented on storage/tokudb/ha_tokupart.h:
```diff
@@ -26,15 +26,11 @@ class ha_tokupart : public native_part::Partition_base {
+                TABLE_SHARE *table_arg,
+                partition_info *part_info)
+        : native_part::Partition_base(hton, table_arg, &table_arg->ha_share) {
+        this->get_partition_handler()->set_part_info(part_info, true);
```
Likewise

satya-bodapati commented:
@laurynas-biveinis replied to your question

satya-bodapati commented:
@laurynas-biveinis Thanks for the review! :+1:
